xaitk_saliency/impls/vis_sal_similarity/similarityscoring.py

```diff
@@ -0,0 +1,61 @@
+from xaitk_saliency import ImageSimilaritySaliencyMapGenerator
+from xaitk_saliency.utils.masking import weight_regions_by_scalar
+
+import numpy as np
+from sklearn.preprocessing import minmax_scale
+from sklearn.metrics.pairwise import euclidean_distances
+
+
+class SimilarityScoring (ImageSimilaritySaliencyMapGenerator):
+    """
+    This saliency implementation transforms proximity in feature
+    space into saliency heatmaps. This should
+    require a sequence of feature vectors of the query and
+    reference image, a number of feature vectors as predicted
+    on perturbed images, as well as the masks of the reference image
+    perturbations (as would be output from a
+    `PerturbImage` implementation.
+
+    The perturbation masks used by the following implementation are
+    expected to be of type integer. Masks containing values of type
+    float are rounded to the nearest value and binarized
+    with value 1 replacing values greater than or equal to half of
+    the maximum value in mask after rounding while 0 replaces the rest.
+    """
+
+    def generate(
+        self,
+        ref_descr_1: np.ndarray,
+        ref_descr_2: np.ndarray,
+        perturbed_descrs: np.ndarray,
+        perturbed_masks: np.ndarray,
+    ) -> np.ndarray:
+
+        # Computing original proximity between image1 and image2 feature vectors.
+        original_proximity = euclidean_distances(ref_descr_1.reshape(1, -1), ref_descr_2.reshape(1, -1))
+
+        # Computing proximity between original image1 and perturbed image2 feature vectors.
+        perturbed_proximity = euclidean_distances(ref_descr_1.reshape(1, -1), perturbed_descrs)[0]
+
+        if len(perturbed_proximity) != len(perturbed_masks):
+            raise ValueError("Number of perturbation masks and respective",
+                             "confidence lengths do not match.")
+
+        # Iterating through each distance and compare it with
+        # its perturbed twin
+        diff_abs = abs(perturbed_proximity - original_proximity)
+
+        diff = np.transpose(np.clip(diff_abs, a_min=0, a_max=np.max(diff_abs)))
```

Comment:
Is a_max necessary here? np.clip documentation says if a_max is None, no clipping is done on the max value.

Reply:
Changed a_max to None.

Reply:
Really minor point, but I would recommend the following for clarity:

```python
diff = np.transpose(np.clip(diff_abs, a_min=0))
```

We could also potentially rename `diff_abs` for clarity, since we removed the previous `abs` operation.

Reply:
Done.